heron/packing/src/java/org/apache/heron/packing/constraints/InstanceDensityConstraint.java

```diff
@@ -0,0 +1,48 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.heron.packing.constraints;
+
+import org.apache.heron.packing.builder.Container;
+import org.apache.heron.packing.exceptions.TooManyInstancesException;
+import org.apache.heron.spi.packing.PackingPlan;
+
+/**
+ * Constraint on the number of instances in one container
+ */
+public class InstanceDensityConstraint implements PackingConstraint {
+  private final int maxNumInstancesPerContainer;
+
+  public InstanceDensityConstraint(int maxNumInstancesPerContainer) {
+    this.maxNumInstancesPerContainer = maxNumInstancesPerContainer;
+  }
+
+  @Override
+  public void test(Container container, PackingPlan.InstancePlan instancePlan)
```

Comment:
maybe rename to "validate"?

Reply:
Sure. I named it `test` because I first wanted to use `Predicate` for this, but it's fine.